src/storage/transaction/TransactionManager.cpp

```diff
@@ -0,0 +1,506 @@
+                *spPromiseVal = nebula::error(errOrVal);
+                return;
+            }
+            auto lockedPtr = result->wlock();
+            if (!lockedPtr) {
+                *spPromiseVal = cpp2::ErrorCode::E_MUTATE_EDGE_CONFLICT;
+                return;
+            }
+            auto& kv = *lockedPtr;
+            if (FLAGS_enable_multi_versions) {
+                auto verPrev =
+                    NebulaKeyUtils::isEdge(vIdLen, kv.first)
+                        ? folly::Endian::big(NebulaKeyUtils::getVersion(vIdLen, kv.first))
+                        : std::numeric_limits<int64_t>::max();
+                auto verCurr = folly::Endian::big(ver);
+                LOG_IF(INFO, FLAGS_trace_toss)
+                    << "compare ver=" << ver << ", verCurr=" << verCurr << ", verPrev=" << verPrev;
+                if (verCurr > verPrev) {
+                    *spPromiseVal = cpp2::ErrorCode::E_MUTATE_EDGE_CONFLICT;
+                    return;
+                }
+            }
+            kv.first = localKey;
+            kv.second = nebula::value(errOrVal);
+            // 3rd, commit local key(indexes)
+            auto localPart = NebulaKeyUtils::getPart(localKey);
+            if (FLAGS_enable_multi_versions) {
+                commitBatch(spaceId, localPart, std::move(lockVal))
+                    .via(exec_.get())
+                    .thenValue([=](auto&& rc) { *spPromiseVal = CommonUtils::to(rc); })
+                    .thenError([=](auto&&) { *spPromiseVal = cpp2::ErrorCode::E_UNKNOWN; });
+            } else {
+                auto key = kv.first;
+                auto val = kv.second;
+                commitEdgeOut(vIdLen, spaceId, localPart, std::move(kv.first), std::move(kv.second))
+                    .via(exec_.get())
+                    .thenValue([=](auto&& rc) { *spPromiseVal = CommonUtils::to(rc); })
+                    .thenError([=](auto&&) { *spPromiseVal = cpp2::ErrorCode::E_UNKNOWN; });
+            }
+        })
+        .thenValue([=](auto&&) {
+            // 4th, remove persist lock
+            if (*spPromiseVal == cpp2::ErrorCode::SUCCEEDED) {
+                erasePersistLock(vIdLen, spaceId, localKey)
+                    .via(exec_.get())
+                    .thenValue([=](auto&& code) { *spPromiseVal = code; })
+                    .thenError([=](auto&&) { *spPromiseVal = cpp2::ErrorCode::E_UNKNOWN; });
+            }
+        })
+        .thenError([=](auto&& ex) {
+            LOG(ERROR) << ex.what();
+            *spPromiseVal = cpp2::ErrorCode::E_UNKNOWN;
+        })
+        .ensure([=, p = std::move(c.first)]() mutable {
+            eraseMemoryLock(localKey, ver);
+            p.setValue(*spPromiseVal);
+        });
+    return std::move(c.second).via(exec_.get());
+}
+
+// combine multi put and remove in a batch
+// this may sometimes reduce some raft operation
+folly::SemiFuture<kvstore::ResultCode> TransactionManager::commitBatch(GraphSpaceID spaceId,
```

Comment:
If I read correctly, all place using `commitBatch` has miss the index

Reply:
goot catch, indexes will added at processor, as a batch

Reply:
My mistake, never mind.